controllers/config/config.go

```diff
@@ -68,12 +68,94 @@ func fillConfig(clusterConfig *configv1.Network, operConfig *operatorv1.AntreaIn
 	return nil
 }
 
+func fillControllerConfig(clusterConfig *configv1.Network, operConfig *operatorv1.AntreaInstall) error {
+	var controllerConfig ctlconfig.ControllerConfig
+	err := yaml.Unmarshal([]byte(operConfig.Spec.AntreaControllerConfig), &controllerConfig)
+	if err != nil {
+		return fmt.Errorf("failed to parse AntreaControllerConfig: %v", err)
+	}
+
+	//Turn on NodeIPAM featureGate.
+	if controllerConfig.FeatureGates == nil {
+		controllerConfig.FeatureGates = make(map[string]bool)
+	}
+	controllerConfig.FeatureGates[string(features.NodeIPAM)] = true
+	controllerConfig.NodeIPAM.EnableNodeIPAM = true
+
+	if clusterConfig != nil {
+		ip4found := false
+		ip6found := false
+
+		for _, cidr := range clusterConfig.Spec.ClusterNetwork {
+			if net.IsIPv4CIDRString(cidr.CIDR) {
+				if !ip4found {
+					ip4found = true
+					controllerConfig.NodeIPAM.ClusterCIDRs = append(controllerConfig.NodeIPAM.ClusterCIDRs, cidr.CIDR)
+					controllerConfig.NodeIPAM.NodeCIDRMaskSizeIPv4 = int(cidr.HostPrefix)
+				}
+			} else {
+				if !ip6found {
+					ip6found = true
+					controllerConfig.NodeIPAM.ClusterCIDRs = append(controllerConfig.NodeIPAM.ClusterCIDRs, cidr.CIDR)
+					controllerConfig.NodeIPAM.NodeCIDRMaskSizeIPv6 = int(cidr.HostPrefix)
+				}
+			}
+		}
+
+		// Set service CIDR
+		ip4found = false
+		ip6found = false
+
+		for _, svcCIDR := range clusterConfig.Spec.ServiceNetwork {
+			if net.IsIPv4CIDRString(svcCIDR) {
+				if !ip4found {
+					ip4found = true
+					controllerConfig.NodeIPAM.ServiceCIDR = svcCIDR
+				}
+			} else {
+				if !ip6found {
+					ip6found = true
+					controllerConfig.NodeIPAM.ServiceCIDRv6 = svcCIDR
+				}
+			}
+		}
+	}
+
+	updatedAntreaControllerConfig, err := yaml.Marshal(controllerConfig)
```

Comment:
Probably still me being obtuse, but I'm not sure why we need to copy structs from Antrea from marshalling and unmarhsalling. I think the existing fillConfig code makes already changes that ultmately end up in Antrea agent config.

Reply:
This is basically how it's done in #47 - not that the controller config and NodeIPAM within is hierarchical and therefore is more cumbersome to implement with maps of interfaces.
And the more params we want to set within the configs, the worse it becomes (and requires more effort to revert).

Reply:
I figured out the additional challenge after looking at the code in PR #47. 
I see the implementation here is more straightforward. The downside is about having the additional hassle of periodically syncrhonizing structs for Antrea Config.

Reply:
Maybe if we choose to take this approach, it's worthwhile to add a periodic CI job which validates the config structs against Antrea.